go/consensus/tendermint/apps/beacon/state/state.go

```diff
@@ -37,53 +37,58 @@ func NewImmutableState(state abci.ApplicationState, version int64) (*ImmutableSt
 }
 
 // Beacon gets the current random beacon value.
-func (s *ImmutableState) Beacon() ([]byte, error) {
-	_, b := s.Snapshot.Get(beaconKeyFmt.Encode())
-	if b == nil {
+func (s *ImmutableState) Beacon(ctx context.Context) ([]byte, error) {
+	data, err := s.Tree.Get(ctx, beaconKeyFmt.Encode())
+	if err != nil {
+		return nil, abciAPI.UnavailableStateError(err)
+	}
+	if data == nil {
 		return nil, beacon.ErrBeaconNotAvailable
 	}
-
-	return b, nil
+	return data, nil
 }
 
-func (s *ImmutableState) ConsensusParameters() (*beacon.ConsensusParameters, error) {
-	_, raw := s.Snapshot.Get(parametersKeyFmt.Encode())
-	if raw == nil {
+func (s *ImmutableState) ConsensusParameters(ctx context.Context) (*beacon.ConsensusParameters, error) {
+	data, err := s.Tree.Get(ctx, parametersKeyFmt.Encode())
+	if err != nil {
+		return nil, abciAPI.UnavailableStateError(err)
+	}
+	if data == nil {
 		return nil, errors.New("tendermint/beacon: expected consensus parameters to be present in app state")
 	}
 
 	var params beacon.ConsensusParameters
-	err := cbor.Unmarshal(raw, &params)
-	return &params, err
+	if err = cbor.Unmarshal(data, &params); err != nil {
+		return nil, abciAPI.UnavailableStateError(err)
```

Comment:
Yeah the assumption here is that if unmarshal fails then it must be due to some kind of data corruption which may be non-deterministic. Of course it could also be deterministic in case of bugs that deterministically corrupt state, but you can't know.

Reply:
oh